### PR TITLE
feat: support opening source files by clicking them on macOS

### LINKED ISSUE #846: Associate source files to open in cpeditor on macOS
**Is your feature request related to a problem? Please describe.**

CPEditor doesn't register itself as a cpp file open method , so I can't open cpp files via CPEditor in my file manager.

**Describe the solution you'd like**

Register CPEditor as a cpp file open method.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by swift-zym**
## Description

Listen `QFileOpenEvent` to open files and register as an editor for code files on macOS.
## Related Issues / Pull Requests

Closes #846 

## Screenshots (if appropriate)

Untitled

## Checklist

- [x] If the key of a setting is changed, the `old` attribute is updated or it is resolved in SettingsUpdater.
- [x] If there are changes of the text displayed in the UI, they are wrapped in `tr()` or `QCoreApplication::translate()`.
- [x] If needed, I have opened a pull request or an issue to update the [documentation](https://github.com/cpeditor/cpeditor.github.io).
- [ ] If these changes are notable, they are documented in [CHANGELOG.md](https://github.com/cpeditor/cpeditor/blob/master/CHANGELOG.md).

### CODE REVIEW

File: src/application.hpp
Comment (sonulohani) [inline src/application.hpp:27]: Use explicit before ctor. Move the implementation to cpp file.

File: src/appwindow.hpp
Comment (sonulohani) [inline src/appwindow.hpp:60]: Dont use friend class. This will break the oops principles. Use setter getter if you want any attribute access from the class

File: src/application.hpp
Comment (sonulohani) [inline src/application.hpp:24]: Whats the intention of bringing new class. Can't you use existing SingleApplication class?

File: src/application.cpp
Comment (sonulohani) [inline src/application.cpp:32]: Should be "Opening file"

File: src/application.cpp
Comment (sonulohani) [inline src/application.cpp:36]: QApplication is singleton class. Just use `QApplication::instance()` to get the QApplication object. No need to iterate over all the objects.

**Review by sonulohani — CHANGES_REQUESTED** [review]

File: src/application.hpp
Comment (coder3101) [reply src/application.hpp]: explicit is required for constructors taking only 1 non default parameter

See:  https://en.cppreference.com/w/cpp/language/explicit

File: src/application.hpp
Comment (coder3101) [reply src/application.hpp:24]: How to Override event method without inheriting?

Comment (sonulohani) [reply src/application.hpp:24]: > How to Override event method without inheriting?

Are you asking me?

Comment (coder3101) [reply src/application.hpp:24]: I mean there was no other way around except inherit to override

Comment (coder3101) [reply src/application.hpp:24]: See this: https://doc.qt.io/qt-5/qfileopenevent.html

Comment (sonulohani) [reply src/application.hpp:24]: I didnt know that there is no singleapplication.hpp file. I thought we could reuse SingleApplication class. 😀😀😀

**Comment by coder3101** [comment]
From QFileOpenEvent docs

> Note: This class is currently supported for macOS only.

I expected some OS macros, how is it being handled in other OS?

File: src/application.hpp
Comment (sonulohani) [reply src/application.hpp]: that is until c++11. Its always good practice to have explicit before ctor

Comment (coder3101) [reply src/application.hpp]: If static analysis passed it's okay.

File: src/application.cpp
Comment (coder3101) [reply src/application.cpp:32]:
```suggestion
        LOG_INFO("Opening file : " << file);
```

File: src/application.cpp
Comment (coder3101) [reply src/application.cpp:36]: `openTab` is a member of applicationwindow.

**Review by sonulohani — DISMISSED** [review]
LGTM. Wait for others also to review the code before merging it into master.

**Review by ouuan — APPROVED** [review]
Now it looks good to me. However, we still need someone to test it on macOS.

File: src/application.cpp
Comment (ouuan) [inline src/application.cpp:29]: Do we really need `#ifdef Q_OS_MACOS`? This class currently supports macOS only, but it's harmless on unsupported platforms.

Comment (ouuan) [reply src/application.cpp:29]: BTW, `Q_OS_MAC` is deprecated, and we should use `Q_OS_MACOS`.

Comment (swift-zym) [reply src/application.cpp:29]: I can't expect what will this class do on other OSs,so I added these macros.

Comment (ouuan) [reply src/application.cpp:29]: > I can't expect what will this class do on other OSs

It should do the same thing. When it's unsupported, I think this event won't be triggered.

**Comment by coder3101** [comment]
> Now it looks good to me. However, we still need someone to test it on macOS.

I will test and let you know.

**Review by ouuan — APPROVED** [review]

**Review by coder3101 — APPROVED** [review]
Looks good and works fine as well.

**Comment by swift-zym** [comment]
BTW, should we also do this for Windows (by editing the registry files)?

**Comment by ouuan** [comment]
> BTW, should we also do this for Windows (by editing the registry files)?

Oh, I forgot that this is just a part of #846. We can do it for both Windows and Linux.

**Comment by ouuan** [comment]
I think we'd better open a new issue. #846 was two problems mixed into one issue.

**Comment by swift-zym** [comment]
> I think we'd better open a new issue. #846 was two problems mixed into one issue.

I've open #855.

**Pull request merged**
